lib/livebook_web/live/session_live/secrets_component.ex

```diff
@@ -273,40 +307,52 @@ defmodule LivebookWeb.SessionLive.SecretsComponent do
   end
 
   defp prefill_secret_name(socket) do
-    if unavailable_secret?(
-         socket.assigns.prefill_secret_name,
-         socket.assigns.secrets,
-         socket.assigns.livebook_secrets
-       ),
-       do: socket.assigns.prefill_secret_name,
-       else: ""
+    if unavailable_secret?(socket, socket.assigns.prefill_secret_name),
+      do: socket.assigns.prefill_secret_name,
+      else: ""
   end
 
-  defp unavailable_secret?(nil, _, _), do: false
-  defp unavailable_secret?("", _, _), do: false
+  defp unavailable_secret?(_socket, nil), do: false
+  defp unavailable_secret?(_socket, ""), do: false
 
-  defp unavailable_secret?(preselect_name, secrets, livebook_secrets) do
-    not Map.has_key?(secrets, preselect_name) and
-      not Map.has_key?(livebook_secrets, preselect_name)
+  defp unavailable_secret?(socket, preselect_name) do
+    not session?(socket, preselect_name) and
+      not livebook?(socket, preselect_name) and
+      not hub?(socket, preselect_name)
   end
 
   defp title(%{assigns: %{select_secret_ref: nil}}), do: "Add secret"
   defp title(%{assigns: %{select_secret_options: %{"title" => title}}}), do: title
   defp title(_), do: "Select secret"
 
-  defp set_secret(socket, secret, "session") do
+  defp build_attrs(%{
+         "name" => name,
+         "value" => value,
+         "store" => "hub",
+         "connected_hub" => hub_id
+       }) do
+    %{name: name, value: value, origin: {:hub, hub_id}}
+  end
+
+  defp build_attrs(%{"name" => name, "value" => value, "store" => "livebook"}) do
+    %{name: name, value: value, origin: :app}
+  end
+
+  defp build_attrs(%{"name" => name, "value" => value, "store" => "session"}) do
+    %{name: name, value: value, origin: :session}
+  end
+
+  defp set_secret(socket, %Secret{origin: :session} = secret) do
     Livebook.Session.set_secret(socket.assigns.session.pid, secret)
   end
 
-  defp set_secret(socket, secret, "livebook") do
+  defp set_secret(socket, %Secret{origin: :app} = secret) do
     Livebook.Secrets.set_secret(secret)
     Livebook.Session.set_secret(socket.assigns.session.pid, secret)
   end
 
-  defp set_secret(socket, secret, "hub") do
-    selected_hub = socket.assigns.data["connected_hub"]
-
-    if hub = Enum.find(socket.assigns.connected_hubs, &(&1.hub.id == selected_hub)) do
+  defp set_secret(socket, %Secret{origin: {:hub, id}} = secret) when is_binary(id) do
+    if hub = Enum.find(socket.assigns.connected_hubs, &(&1.hub.id == id)) do
       create_secret_request =
         LivebookProto.CreateSecretRequest.new!(
```

Comment:
Sorry, but let's build the request inside EnterpriseClient. The API should be: `EnterpriseClient.create_secret(name, value)`. The proto files should never leak outside of EnterpriseClient. Feel free to either open an issue to solve this later OR tackle it in this pull request. :)

Reply:
The next PR introduces the `Hubs.Provider.create_secret/2`, which will replace this implementation.